src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs

```diff
@@ -0,0 +1,266 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Core.Caching
+{
+    using NonBlocking;
+
+    public sealed class NonBlockingLruCache<TKey, TValue> : IThreadPoolWorkItem, ICache<TKey, TValue> where TKey : notnull
```

Comment:
I see how you split implementations now. Why not to hide them behind `LruCache` generic class and return one depending on the `Unsafe.SizeOf<TKey>` ?